### PR TITLE
Fix ignored domain events in virt-handler and migration abort test

### PR DESCRIPTION
**Pull request by vasiliy-ul**
**What this PR does / why we need it**:

The first commit addresses the issue wit domain events: previously the callbacks used by the domain informer were not adjusting the VMI expectations when queueing objects. As a result some of the events were ignored by virt-handler. That was causing issues with timely updates of VMI status fields (e.g. the secondary NIC address info reported by Guest Agent was not reflected). Also LowerExpectations was called incorrectly in deleteFunc. Switched to using {Creation,Deletion}Observed helpers.

The second commit fixes the migration abort tests. Sometimes the `MigrationAbortInProgress` status was set *after* `MigrationAbortSuceeded`. The additional check should prevent that scenario and fix flaky test:

> Tests Suite: [Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system][sig-compute] VM Live Migration Starting a VirtualMachineInstance live migration cancelation should be able to cancel a migration [sig-storage][test_id:2226] with ContainerDisk

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Hit the issues while working on https://github.com/kubevirt/kubevirt/pull/6958. Some tests were very flaky.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
/lgtm
Thanks @vasiliy-ul !

**Review by marceloamaral — COMMENTED** [review]
Good catch!

Since you are updating the vmi expectation, you can improve the code with some refactoring.

- When deleting the expectation, it should be `vmiExpectations.LowerExpectations (key, 0, 1)` and not `vmiExpectations.LowerExpectations (key, 1, 0)` ... as [here](https://github.com/kubernetes/kubernetes/blob/885961cda8875cce192700fe6a32ff57ba706106/pkg/controller/controller_utils.go#L260)

- Also, to follow the kubernetes convention and avoid the above described issue, instead of `SetExpectations()` we should use `CreationObserved()` and `DeletionObserved()`

- Finally, I'm not sure if update should be defined in the controller's expectations, we can see that the interface only defines helper functions for ExpectCreation or ExpectDeletion, but not for Updates. Also, I didn't find changes in expectation when updating in any kubernetes controller. So, unless you see an important use case for that, I suggest to remove this.

**Comment by vasiliy-ul** [comment]
> * When deleting the expectation, it should be `vmiExpectations.LowerExpectations (key, 0, 1)` and not `vmiExpectations.LowerExpectations (key, 1, 0)` ... as [here](https://github.com/kubernetes/kubernetes/blob/885961cda8875cce192700fe6a32ff57ba706106/pkg/controller/controller_utils.go#L260)

Wow, good catch. I missed that. Fixed :+1: 

> * Also, to follow the kubernetes convention and avoid the above described issue, instead of `SetExpectations()` we should use `CreationObserved()` and `DeletionObserved()`

Makes sense.

> * Finally, I'm not sure if update should be defined in the controller's expectations, we can see that the interface only defines helper functions for ExpectCreation or ExpectDeletion, but not for Updates. Also, I didn't find changes in expectation when updating in any kubernetes controller. So, unless you see an important use case for that, I suggest to remove this.

Actually the issue I was trying to fix was related to updates. In one of the tests the secondary NIC address reported by the Guest Agent was reflected in the VMI status only after ~5min (causing a timeout and test failure). I traced the event up to the `updateDomainFunc` and found that it was dropped at:

https://github.com/kubevirt/kubevirt/blob/f8160a0b5eac2acafab74fd7916bfcaf0c39a6a3/pkg/virt-handler/vm.go#L1691-L1693

Then I just checked the git history and it looked like the adjustment of the expectations for the update case was droped accidentally when those `*DomainFunc` were introduced.

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -2638,14 +2638,14 @@ func (d *VirtualMachineController) calculateVmPhaseForStatusReason(domain *api.D
...
-		d.vmiExpectations.LowerExpectations(key, 1, 0)
+		d.vmiExpectations.CreationObserved(key)
 		d.Queue.Add(key)
 	}
 }
 func (d *VirtualMachineController) deleteFunc(obj interface{}) {
 	key, err := controller.KeyFunc(obj)
 	if err == nil {
-		d.vmiExpectations.LowerExpectations(key, 1, 0)
+		d.vmiExpectations.ControllerExpectationsInterface.DeletionObserved(key)
```
Comment (davidvossel) [inline pkg/virt-handler/vm.go:2648]: I know this is weird, but we aren't using expectations here to determine if a VMI has been created or deleted, we're using expectations to determine when a change to the VMI has been received by the informer.

So this is technically correct the way was previously. We aren't detecting creations or deletions, just that the VMI changed.

File: pkg/virt-handler/vm.go [hunk 2]
```diff
@@ -2660,10 +2660,7 @@ func (d *VirtualMachineController) updateFunc(_, new interface{}) {
 func (d *VirtualMachineController) addDomainFunc(obj interface{}) {
 	domain := obj.(*api.Domain)
 	log.Log.Object(domain).Infof("Domain is in state %s reason %s", domain.Status.Status, domain.Status.Reason)
-	key, err := controller.KeyFunc(obj)
-	if err == nil {
-		d.Queue.Add(key)
-	}
+	d.addFunc(obj)
```
Comment (davidvossel) [inline pkg/virt-handler/vm.go:2663]: we don't want a change to the domain to indicate that the VMI informer has been updated (lower expectations). So this was also correct the previous way where `d.addFunc` was not used.

File: pkg/virt-handler/vm.go [hunk 3]
```diff
@@ -2680,10 +2677,7 @@ func (d *VirtualMachineController) deleteDomainFunc(obj interface{}) {
 		}
 	}
 	log.Log.Object(domain).Info("Domain deleted")
-	key, err := controller.KeyFunc(obj)
-	if err == nil {
-		d.Queue.Add(key)
-	}
+	d.deleteFunc(obj)
```
Comment (davidvossel) [inline pkg/virt-handler/vm.go:2680]: this also needs to be reverted

File: pkg/virt-handler/vm.go [hunk 4]
```diff
@@ -2696,10 +2690,7 @@ func (d *VirtualMachineController) updateDomainFunc(old, new interface{}) {
 		log.Log.Object(newDomain).Info("Domain is marked for deletion")
 	}
 
-	key, err := controller.KeyFunc(new)
-	if err == nil {
-		d.Queue.Add(key)
-	}
+	d.updateFunc(old, new)
```
Comment (davidvossel) [inline pkg/virt-handler/vm.go:2693]: this also needs to be reverted

File: pkg/virt-launcher/virtwrap/live-migration-source.go [hunk 5]
```diff
@@ -355,7 +355,7 @@ func (l *LibvirtDomainManager) setMigrationResultHelper(vmi *v1.VirtualMachineIn
 	now := metav1.Now()
 	if abortStatus != "" {
 		metaAbortStatus := domainSpec.Metadata.KubeVirt.Migration.AbortStatus
-		if metaAbortStatus == string(abortStatus) && metaAbortStatus == string(v1.MigrationAbortInProgress) {
+		if metaAbortStatus == string(abortStatus) && metaAbortStatus == string(v1.MigrationAbortInProgress) || metaAbortStatus == string(v1.MigrationAbortSucceeded) {
 			return domainerrors.MigrationAbortInProgressError
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/live-migration-source.go:359]: a migration that has succeeded isn't in progress anymore. I'm unsure about this change.

**Review by davidvossel — COMMENTED** [review]
> The first commit addresses the issue wit domain events: previously the callbacks used by the domain informer were not adjusting the VMI expectations when queueing objects. As a result some of the events were ignored by virt-handler. That was causing issues with timely updates of VMI status fields (e.g. the secondary NIC address info reported by Guest Agent was not reflected). Also LowerExpectations was called incorrectly in deleteFunc. Switched to using {Creation,Deletion}Observed helpers.

That's not how the expectations are intended to work for virt-handler. In virt-handler we aren't creating or deleting VMI objects. The expectations are used to gate when a VMI can be reconciled again after writing a change to the VMI to etcd.

The issue we were having was that we'd write a change to a VMI, then a bunch of domain events would come in triggering new reconciles before the VMI informer caught up. The result when that occurs is that when we try to update the VMI again, that update will fail due to the VMI being outdated (informer didn't have time to catch up yet so we were using an old vmi)

so, the expectation logic is meant as a optimization to avoid API calls that will for sure fail.

Now, all that said. You're experiencing something here that is kind of confusing. It's unclear how the changes you've made here somehow improve the timely update of the VMI status.

**Comment by davidvossel** [comment]
> Actually the issue I was trying to fix was related to updates. In one of the tests the secondary NIC address reported by the Guest Agent was reflected in the VMI status only after ~5min (causing a timeout and test failure).

That's interesting. looking at the code, all the expectation logic seems to line up correctly. this is definitely curious.

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (marceloamaral) [reply pkg/virt-handler/vm.go:2648]: Interesting, this wasn't clear, we should have this commented out in the code to make it easier to follow

**Comment by vasiliy-ul** [comment]
@davidvossel , thank you for bringing clarity on this. Seems that I misunderstood the logic with expectations. Probably the PR does not make much sense then. Though now I am a bit confused about the handling of domain update events. From what I observed, the event with address information for the secondary NIC was received in virt-handler by the corresponding callback `updateDomainFunc`, put into the queue, but then discarded in `func (d *VirtualMachineController) execute(key string) error`:

https://github.com/kubevirt/kubevirt/blob/f8160a0b5eac2acafab74fd7916bfcaf0c39a6a3/pkg/virt-handler/vm.go#L1691-L1693

Perhaps I need to debug the issue further. I think I will close the PR for now. Will reopen it or create a new one when have some updates.

**Comment by davidvossel** [comment]
> Perhaps I need to debug the issue further. I think I will close the PR for now. Will reopen it or create a new one when have some updates.

@vasiliy-ul Here's a thought. The expectation logic in virt-handler is an optimization, and not absolutely critical for operation. For the most part It makes the reconcile loop's usage of the api server more efficient by reducing Update collisions.

I've looked this code over several times, and I can't pick apart a reason for while this optimization might cause a significant delay in updating the VMI. However I don't doubt you're experiencing something here.

As part of your debug, I'd recommend disabling the `d.vmiExpectations.SatisfiedExpectations(key)` in the execute function. If you can prove that's related to the delay you're experiencing somehow, i'd consider removing this optimization entirely.

**Comment by vasiliy-ul** [comment]
@davidvossel , hey, thanks for the hints on where to look at. I still do not exclude the fact that I might have messed up smth :)

I haven't yet restored the environment to do some more in-depth testing... Though I may summarize how I traced the issue. Maybe you could have a look and perhaps tell if smth is incorrect in the flow:

* The event with interfaces update is fired up by the `agent-poller` in `virt-launcher` and eventually it gets into `eventCallback` here:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-launcher/notify-client/client.go#L421-L430

* and then into `client.SendDomainEvent(watch.Event{Type: watch.Modified, Object: domain})`:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-launcher/notify-client/client.go#L323-L349

* At this point it is sent via IPC to `virt-handler`'s `HandleDomainEvent`:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-handler/notify-server/server.go#L82-L83

* and then propagated via the `EventChan` to the `updateDomainFunc` where it is put into the work queue:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-handler/vm.go#L2720-L2723

* After that it is picked up by the worker thread and `execute` is called:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-handler/vm.go#L1704

* Where the event gets ignored since there is a check for expectations which have not been adjusted:

https://github.com/kubevirt/kubevirt/blob/108402380b8c2ecef8dde8f0cf3319e98f3331d2/pkg/virt-handler/vm.go#L1712-L1713


So basically that is what I observed. Is it the expected behaviour? In theory even if the domain event is discarded its not smth critical. It will just be handled later along with some other event. But who guarantees the next event comes in some reasonable ammount of time? I would assume there is a periodic timer-based check for that?

**Comment by vasiliy-ul** [comment]
UPD: so tested again with the latest master and can observe the same. Unless I am doing smth completely wrong, the flow described above should be valid.

The domain update events are discarded in this check `} else if !d.vmiExpectations.SatisfiedExpectations(key) {` with the message `Controller (name: n/a) still waiting on expectations \u0026controller.ControlleeExpectations{add:1, del:0, key:\"kubevirt-test-default1/testvmi-vnmz7\`. 

The issue can be fixed either by adjusting the expectations `d.vmiExpectations.LowerExpectations(key, 1, 0)` in `updateDomainFunc` or by removing the `SatisfiedExpectations` check.

**Comment by davidvossel** [comment]
> The issue can be fixed either by adjusting the expectations d.vmiExpectations.LowerExpectations(key, 1, 0) in updateDomainFunc or by removing the SatisfiedExpectations check.

wow, that is certainly confusing. I've looked this expectation code over and i don't have a great explanation for how this is possible. The expectation logic is supposed to ensure that we only process the vmi again after we've seen the results from the latest Update/Patch. The domain events don't have anything to do with this really. It's all about getting the vmi informer synced before reconciling the VMI.


Do you have modifications in your code? If not i think we should remove this expectation logic entirely.

One possibility here is that the expectation logic gets out of sync due to the handoff we perform from one handler to another. If you are making no modifications and this is occurring with a clean kubevirt release, then I think we should remove the expectation logic from virt-handler entirely.

**Comment by vasiliy-ul** [comment]
> wow, that is certainly confusing. I've looked this expectation code over and i don't have a great explanation for how this is possible. The expectation logic is supposed to ensure that we only process the vmi again after we've seen the results from the latest Update/Patch. The domain events don't have anything to do with this really. It's all about getting the vmi informer synced before reconciling the VMI.

Now I am confused even more :) The code that updates the VMI and its status is further in the flow: `execute(key string) error` -> `defaultExecute(key string, ...) error` -> `updateVMIStatus(...) error` and its not reached since the expectation check performed in `execute` makes it return immediately. So the status gets updated later, only when some other event which lowers the expectations is triggered. If this is the expected behaviour then I do not understand the need to even monitor for domain updates and put them in the work queue. 


> Do you have modifications in your code? 

Initially I faced that issue while working on another PR. But then I also checked with the clean master and it was the same. Apart from some additional debug prints there were no code changes. I hit the issue with this test:

https://github.com/kubevirt/kubevirt/blob/677b9e82d44e88a0dbc6486bb6562b41ec93b32d/tests/network/vmi_multus.go#L1494

The failue happens in `BeforeEach` where a Fedora VMI is created and it should report the IP address of the secondary NIC in its status. There is a 4min timeout which sometimes appears to be enough though.

> One possibility here is that the expectation logic gets out of sync due to the handoff we perform from one handler to another. 

Hm... what handoff? Do you mean somehow the expectations are modified in the time period after a domain update event is added to the queue but not yet processed?

**Comment by vasiliy-ul** [comment]
UPD: Hm... Strange, but I do not see that issue anymore with the latest master. Though I was able to reproduce it a week ago... now I am totally lost :confused:

**Comment by davidvossel** [comment]
> Hm... what handoff? Do you mean somehow the expectations are modified in the time period after a domain update event is added to the queue but not yet processed?

during a live migration, the handler on the source node hands the VMI off to the handler on the destination node by modifying a label on the VMI. That results in the informer on the source node never seeing the result of that last update that performs the handoff. 

> UPD: Hm... Strange, but I do not see that issue anymore with the latest master. Though I was able to reproduce it a week ago... now I am totally lost confused

weird... I'm still not convinced this isn't a real issue. Do you have any other thoughts on how to reproduce? Or are you interested in giving this any more time? 

I made a PR #7126 to to remove the optimization that is related to your timing issue. I just want to make sure this is reproducible in an environment that is running unmodified main branch or release branch code.

**Comment by vasiliy-ul** [comment]
> during a live migration, the handler on the source node hands the VMI off to the handler on the destination node by modifying a label on the VMI. That results in the informer on the source node never seeing the result of that last update that performs the handoff.

Ah, you were refering to live migration flow. Not sure its the case here. The issue was reproducable without the actual migration being scheduled/started.

> weird... I'm still not convinced this isn't a real issue. Do you have any other thoughts on how to reproduce? Or are you interested in giving this any more time?

Yeah, for sure I want to spent more time and experiment with it. Will look at it next week.

> I made a PR #7126 to to remove the optimization that is related to your timing issue.

Great. I will comment there when have some new findings.

> I just want to make sure this is reproducible in an environment that is running unmodified main branch or release branch code.

I will try with 0.49.0 release. Maybe some commit in the main branch has impacted that code flow.

Initially I hit this problem in the PR https://github.com/kubevirt/kubevirt/pull/6958 where I tried to update the fedora-with-test-tooling image to Fedora 35. I was able to fix the test failure either by increasing the timeout for VMI status update up to ~5min or by lowering the expectations. At that point it was reproducable locally with the main branch as well.

Now its a bit wierd. With the latest clean `main` the test does not fail anymore. However after rebasing the PR with Fedora 35, all the network jobs fail again on that exact tests:

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6958/pull-kubevirt-e2e-k8s-1.21-sig-network/1484580324418523136
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6958/pull-kubevirt-e2e-k8s-1.22-sig-network/1484580336842051584
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6958/pull-kubevirt-e2e-k8s-1.23-sig-network/1484580341275430912

So will try to debug it further...

**Comment by EdDev** [comment]
@vasiliy-ul , I think the problem is there on main as well: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7110/pull-kubevirt-e2e-k8s-1.22-sig-network/1485206347384360960

I guess it is a timing thing.

**Comment by vasiliy-ul** [comment]
> @vasiliy-ul , I think the problem is there on main as well: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7110/pull-kubevirt-e2e-k8s-1.22-sig-network/1485206347384360960

Yeah, I now see that it is a flaky issue. I can reproduce it on t`main` and `v0.49.0`. Though some times the test passes normally.

> 
> I guess it is a timing thing.

Well, there is alsready a `4min` timeout. IMHO that should be more than enough. Also I can reproduce it locally (i.e. the 'CI under load' factor does not apply here).

**Comment by vasiliy-ul** [comment]
I think I've found what is causing the issue. Initially I misunderstood the logic with expectations and started looking into the wrong direction. What is actually happening here is that `virt-handler` on certain point calls `Update` on the `VMI` object and sets the expectations. But for some reason the update is never observed with the informer callback.

So I looked at the actual `Update` call. Even with a `DeepEqual` check in the code, `Update` is still invoked when there is **no change** in the VMI object:

https://github.com/kubevirt/kubevirt/blob/500045ed0910c8dd7945d7841e1347a4b7cfb0b2/pkg/virt-handler/vm.go#L1130-L1138

... and therefore no corresponding event is received later on to lower the expectations:

https://github.com/kubevirt/kubevirt/blob/500045ed0910c8dd7945d7841e1347a4b7cfb0b2/pkg/virt-handler/vm.go#L2673-L2679

The problem here is with the equality of the structures `oldStatus` and `vmi.Status`. While `reflect.DeepEqual` considers them as **not equal**, their **json representation is absolutely identical**. That can happen if the compared structs have e.g. slice fields. In this exact case: 
```go
IPs []string `json:"ipAddresses,omitempty"`
```
One of the structures has this field **unitialized** (i.e. `== nil`) while the other one - initialized but **empty** (i.e. `[]string{} != nil`). So from `reflect.DeepEqual` perspective they are different but semantically they are not. Here is a small reproducer: https://go.dev/play/p/4tYECKpBf2Q

I am now thinking how to fix that. This `DeepEqual` check is used in many similar places across the codebase. Probaly it would be better to replace it completely to avoid similar problems in the future. The first thing that came to my mind was to convert the objects to `json` and to compare the bytes/string representations. That actually worked and fixed the failing test.

Then I also found https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go but havent tried this one so not yet sure if its suitable for the usecase here. Will have a closer look...

@davidvossel , @EdDev FYI

**Comment by EdDev** [comment]
Wow, this is a super nasty bug.

A combination of golang letting you feel comfortable with nil/s and empty slices/maps, behaving the same.. almost. And then lacking to cover this in unit test, which seems too important to miss.

Thanks a lot for finding this! We should nominate this one as the bug of the year.

**Comment by vasiliy-ul** [comment]
Yeah, tricky bug...

So I tried with `equality.Semantic.DeepEqual` and it also seems to solve the problem. For now opened a PR with fixing only the VMI controller in virt-handler https://github.com/kubevirt/kubevirt/pull/7131. If the appraoch is fine I can create additional PRs. Though if replacing reflect.DeepEqual through all the code it may affect the tests (and it does actually).

**Comment by rmohr** [comment]
> So I tried with `equality.Semantic.DeepEqual` and it also seems to solve the problem. For now opened a PR with fixing only the VMI controller in virt-handler #7131. If the appraoch is fine I can create additional PRs. Though replacing reflect.DeepEqual through all the code may affect the tests (and it does actually).

I think independent of we take in the PR from David (https://github.com/kubevirt/kubevirt/pull/7126), we want this fixed :) Great that you found it.
